File: computing/interview/design-patterns/structural/composite/python/composite_demo.py
from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List


class FileSystemEntry(ABC):
    def __init__(self, name: str) -> None:
        self._name = name

    @abstractmethod
    def print(self, indent_level: int) -> None:
        ...

    def _indent(self, level: int) -> None:
        print("  " * level, end="")


class FileEntry(FileSystemEntry):
    def print(self, indent_level: int) -> None:
        self._indent(indent_level)
        print(f"File: {self._name}")


class DirectoryEntry(FileSystemEntry):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self._children: List[FileSystemEntry] = []

    def add(self, entry: FileSystemEntry) -> None:
        self._children.append(entry)

    def print(self, indent_level: int) -> None:
        self._indent(indent_level)
        print(f"Dir: {self._name}")
        for child in self._children:
            child.print(indent_level + 1)


def main() -> None:
    root = DirectoryEntry("root")
    docs = DirectoryEntry("docs")
    img = FileEntry("image.png")

    docs.add(FileEntry("resume.pdf"))
    root.add(docs)
    root.add(img)

    root.print(0)


if __name__ == "__main__":
    main()

File: computing/interview/design-patterns/structural/composite/java/CompositeDemo.java
package interview.designpatterns.structural.composite;

import java.util.ArrayList;
import java.util.List;

abstract class FileSystemEntry {
    private final String name;

    FileSystemEntry(String name) {
        this.name = name;
    }

    protected String name() {
        return name;
    }

    abstract void print(int indentLevel);

    protected void indent(int level) {
        for (int i = 0; i < level; i++) {
            System.out.print("  ");
        }
    }
}

final class FileEntry extends FileSystemEntry {
    FileEntry(String name) {
        super(name);
    }

    @Override
    void print(int indentLevel) {
        indent(indentLevel);
        System.out.printf("File: %s%n", name());
    }
}

final class DirectoryEntry extends FileSystemEntry {
    private final List<FileSystemEntry> children = new ArrayList<>();

    DirectoryEntry(String name) {
        super(name);
    }

    void add(FileSystemEntry entry) {
        children.add(entry);
    }

    @Override
    void print(int indentLevel) {
        indent(indentLevel);
        System.out.printf("Dir: %s%n", name());
        for (FileSystemEntry child : children) {
            child.print(indentLevel + 1);
        }
    }
}

public final class CompositeDemo {
    private CompositeDemo() {}

    public static void main(String[] args) {
        DirectoryEntry root = new DirectoryEntry("root");
        DirectoryEntry docs = new DirectoryEntry("docs");
        FileEntry img = new FileEntry("image.png");

        docs.add(new FileEntry("resume.pdf"));
        root.add(docs);
        root.add(img);

        root.print(0);
    }
}
